src/ishikawa_backend.erl

```diff
@@ -101,40 +105,104 @@ init([]) ->
     %% First, increment the vector.
     VClock = vclock:increment(Actor, VClock0),
 
     %% Generate message.
-    Message = {tcbcast, Actor, encode(Message), VClock},
+    Message1 = {tcbcast, Actor, encode(Message), VClock},
 
     %% Transmit to membership.
-    [send(Message, Peer) || Peer <- Members],
+    [send(Message1, Peer) || Peer <- Members],
 
-    {reply, ok, State#state{vv=VClock}};
-handle_call({tcbdeliver, _Message, _Timestamp}, _From, State) ->
-    %% TODO: Implement me.
-    {reply, ok, State};
+    %% Add members to the queue of not ack messages.
+    ToBeAckQueue1 = ToBeAckQueue ++ [{{Actor, VClock}, Members}],
+
+    {reply, ok, State#state{vv=VClock, to_be_ack_queue=ToBeAckQueue1}};
+handle_call({tcbdeliver, Origin, Message, Timestamp}, _From, #state{vv=VClock0,
+                                              rtm=RTM0,
+                                              to_be_delivered_queue=Queue0} = State) ->
+
+    %% Check if the message should be delivered and delivers it or not
+    {VClock, Queue} = trcb:causal_delivery({Origin, Message, Timestamp}, VClock0, Queue0),
+
+    %% Update the Recent Timestamp Matrix
+    RTM = mclock:update_rtm(RTM0, Origin, Timestamp),
+
+    %% Update the Stable Version Vector
+    SVV = mclock:update_stablevv(RTM),
+
+    {reply, ok, State#state{vv=VClock, to_be_delivered_queue=Queue, svv=SVV, rtm=RTM}};
 handle_call({tcbstable, _Timestamp}, _From, State) ->
     %% TODO: Implement me.
     {reply, {ok, false}, State};
-handle_call(Msg, _From, State) ->
-    lager:warning("Unhandled messages: ~p", [Msg]),
-    {reply, ok, State}.
+handle_call({tcbcast, Actor, Message, VClock} = Msg, From, #state{to_be_ack_queue=Queue0, members=Members} = State) ->
+    case lists:keyfind({Actor, VClock}, 1, Queue0) of
+        {_, _} ->
+            %% Generate message.
+            MessageAck = {tcbcast_ack, Actor, Message, VClock},
+
+            Queue1 = Queue0,
+
+            %% Send Ack back to message sender
+            send(MessageAck, From);
+        false ->
+            %% Add members to the queue of not ack messages.
+            Queue1 = Queue0 ++ [{Actor, Message, VClock}, Members],
+
+            %% Transmit to membership.
+            [send(Msg, Peer) || Peer <- Members],
+
+            %% Generate message.
+            MessageAck = {tcbcast_ack, Actor, Message, VClock},
+
+            %% Send Ack back to message sender
+            send(MessageAck, From)
+    end,
+    {reply, ok, State#state{to_be_delivered_queue=Queue1}};
+handle_call({tcbcast_ack, Actor, Message, VClock}, From, #state{to_be_ack_queue=QueueAck0} = State) ->
+    case lists:keyfind({Actor, VClock}, 1, QueueAck0) of
+        {_, QueueMsg} ->
+            case length(QueueMsg)>0 of
+                true ->
+                    case length(QueueMsg) of
+                        1 ->
+                            QueueMsg1 = lists:delete(From, QueueMsg),
```

Comment:
It seems like this could be outside of the case statement, given the behavior is exactly the same for both cases.